includes/Core/Authentication/Google_Proxy.php

```diff
@@ -198,23 +198,45 @@ public function permissions_url( Credentials $credentials, array $query_args = a
 	 * @return string Complete proxy URL.
 	 */
 	public function url( $path = '' ) {
-		$url          = self::PRODUCTION_BASE_URL;
+		$url = self::PRODUCTION_BASE_URL;
+
+		if ( defined( 'GOOGLESITEKIT_PROXY_URL' ) ) {
+			$url = $this->sanitize_base_url( GOOGLESITEKIT_PROXY_URL );
+		}
+
+		$url = untrailingslashit( $url );
+		if ( $path && is_string( $path ) ) {
+			$url .= '/' . ltrim( $path, '/' );
+		}
+
+		return $url;
+	}
+
+	/**
+	 * Sanitizes the given base URL.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @param string $url Base URL to sanitize.
+	 * @return string Sanitized base URL.
+	 */
+	public function sanitize_base_url( $url ) {
 		$allowed_urls = array(
 			self::PRODUCTION_BASE_URL,
 			self::STAGING_BASE_URL,
 			self::DEVELOPMENT_BASE_URL,
 		);
 
-		if ( defined( 'GOOGLESITEKIT_PROXY_URL' ) && in_array( GOOGLESITEKIT_PROXY_URL, $allowed_urls, true ) ) {
-			$url = GOOGLESITEKIT_PROXY_URL;
+		if ( in_array( $url, $allowed_urls, true ) ) {
+			return $url;
 		}
 
-		$url = untrailingslashit( $url );
-		if ( $path && is_string( $path ) ) {
-			$url .= '/' . ltrim( $path, '/' );
+		// Allow for version-specific URLs to application instances.
+		if ( preg_match( '#^https://(?:\d{8}t\d{6}-dot-)?site-kit(?:-dev|-local)?(?:\.[a-z]{2}\.r)?\.appspot\.com/?$#', $url, $_ ) ) {
```

Comment:
This regex technically covers all cases (hence the optional version part) but I kept the above list because it's more straightforward and less likely to be broken accidentally.